File: csharp/Observer/ProductManager.cs
namespace Observer;

public class ProductManager
{
    List<Observer> _observers = new List<Observer>();

    public void UpdatePrice()
    {
        Console.WriteLine("Product price changed.");
        Notify();
    }

    public void Attach(Observer observer)
    {
        _observers.Add(observer);
    }

    public void Detach(Observer observer)
    {
        _observers.Remove(observer);
    }

    private void Notify()
    {
        foreach (Observer observer in _observers)
        {
            observer.Update();
        }
    }
}

File: java/observer/src/main/ProductManager.java
package main;

import java.util.ArrayList;
import java.util.List;

public class ProductManager {

    List<Observer> observers = new ArrayList<>();

    public void updatePrice()
    {
        System.out.println("Product price changed.");
        notification();
    }

    public void attach(Observer observer)
    {
        observers.add(observer);
    }

    public void detach(Observer observer)
    {
        observers.remove(observer);
    }

    private void notification()
    {
        for (Observer observer : observers)
        {
            observer.Update();
        }
    }

}
